File: plugins/wasm-assemblyscript/assembly/request_wrapper.ts
import {
  stream_context,
  log,
  LogLevelValues
} from "@higress/proxy-wasm-assemblyscript-sdk/assembly";

export function getRequestScheme(): string {
  let scheme: string  = stream_context.headers.request.get(":scheme");
  if (scheme == "") {
    log(LogLevelValues.error, "Parse request scheme failed");
  }
  return scheme;
}

export function getRequestHost(): string {
  let host: string = stream_context.headers.request.get(":authority");
  if (host == "") {
    log(LogLevelValues.error, "Parse request host failed");
  }
  return host;
}

export function getRequestPath(): string {
  let path: string = stream_context.headers.request.get(":path");
  if (path == "") {
    log(LogLevelValues.error, "Parse request path failed");
  }
  return path;
}

export function getRequestMethod(): string {
  let method: string = stream_context.headers.request.get(":method");
  if (method == "") {
    log(LogLevelValues.error, "Parse request method failed");
  }
  return method;
}

export function isBinaryRequestBody(): boolean {
  let contentType: string = stream_context.headers.request.get("content-type");
  if (contentType != "" && (contentType.includes("octet-stream") || contentType.includes("grpc"))) {
    return true;
  }

  let encoding: string = stream_context.headers.request.get("content-encoding");
  if (encoding != "") {
    return true;
  }

  return false;
}

export function isBinaryResponseBody(): boolean {
  let contentType: string = stream_context.headers.response.get("content-type");
  if (contentType != "" && (contentType.includes("octet-stream") || contentType.includes("grpc"))) {
    return true;
  }

  let encoding: string = stream_context.headers.response.get("content-encoding");
  if (encoding != "") {
    return true;
  }

  return false;
}

File: plugins/wasm-rust/src/request_wrapper.rs
use proxy_wasm::hostcalls;

use crate::internal;

fn get_request_head(head: &str, log_flag: &str) -> String {
    if let Some(value) = internal::get_http_request_header(head) {
        value
    } else {
        hostcalls::log(
            proxy_wasm::types::LogLevel::Error,
            &format!("get request {} failed", log_flag),
        )
        .unwrap();
        String::new()
    }
}

pub fn get_request_scheme() -> String {
    get_request_head(":scheme", "head")
}

pub fn get_request_host() -> String {
    get_request_head(":authority", "host")
}

pub fn get_request_path() -> String {
    get_request_head(":path", "path")
}

pub fn get_request_method() -> String {
    get_request_head(":method", "method")
}

pub fn is_binary_request_body() -> bool {
    if let Some(content_type) = internal::get_http_request_header("content-type") {
        if content_type.contains("octet-stream") || content_type.contains("grpc") {
            return true;
        }
    }
    if let Some(encoding) = internal::get_http_request_header("content-encoding") {
        if !encoding.is_empty() {
            return true;
        }
    }
    false
}

pub fn is_binary_response_body() -> bool {
    if let Some(content_type) = internal::get_http_response_header("content-type") {
        if content_type.contains("octet-stream") || content_type.contains("grpc") {
            return true;
        }
    }
    if let Some(encoding) = internal::get_http_response_header("content-encoding") {
        if !encoding.is_empty() {
            return true;
        }
    }
    false
}

pub fn has_request_body() -> bool {
    let content_type = internal::get_http_request_header("content-type");
    let content_length_str = internal::get_http_request_header("content-length");
    let transfer_encoding = internal::get_http_request_header("transfer-encoding");
    hostcalls::log(
        proxy_wasm::types::LogLevel::Debug,
        &format!(
            "check has request body: content_type:{:?}, content_length_str:{:?}, transfer_encoding:{:?}",
            content_type, content_length_str, transfer_encoding
        )
    ).unwrap();
    if content_type.is_some_and(|x| !x.is_empty()) {
        return true;
    }
    if let Some(cl) = content_length_str {
        if let Ok(content_length) = cl.parse::<i32>() {
            if content_length > 0 {
                return true;
            }
        }
    }
    transfer_encoding.is_some_and(|x| x == "chunked")
}
